packages/financial-templates-lib/src/clients/FinancialContractFactoryEventClient.js

```diff
@@ -0,0 +1,109 @@
+// A thick client for getting information about FinancialContractFactory events.
+
+class FinancialContractFactoryEventClient {
+  /**
+   * @notice Constructs new FinancialContractEventClient.
+   * @param {Object} logger Winston module used to send logs.
+   * @param {Object} financialContractFactoryAbi ExpiringMultiPartyCreator or PerpetualMultiPartyCreator truffle ABI object.
+   * @param {Object} web3 Web3 provider from truffle instance.
+   * @param {String} financialContractFactoryAddress Ethereum address of the factory contract deployed on the current network.
+   * @param {Integer} startingBlockNumber Offset block number to index events from.
+   * @param {Integer} endingBlockNumber Termination block number to index events until. If not defined runs to `latest`.
+   * @return None or throws an Error.
+   */
+  constructor(
+    logger,
+    financialContractFactoryAbi,
+    web3,
+    financialContractFactoryAddress,
+    startingBlockNumber = 0,
+    endingBlockNumber = null,
+    contractType = "Perpetual"
+    // Default to PerpetualMultiParty for now since the first intended user is the funding rate proposer bot
+  ) {
+    this.logger = logger;
+    this.web3 = web3;
+
+    // Factory contract
+    this.financialContractFactory = new this.web3.eth.Contract(
+      financialContractFactoryAbi,
+      financialContractFactoryAddress
+    );
+    this.financialContractFactoryAddress = financialContractFactoryAddress;
+
+    // Factory Contract Events data structure to enable synchronous retrieval of information.
+    this.createdContractEvents = [];
+
+    // First block number to begin searching for events after.
+    this.firstBlockToSearch = startingBlockNumber;
+
+    // Last block number to end the searching for events at.
+    this.lastBlockToSearchUntil = endingBlockNumber;
+    this.lastUpdateTimestamp = 0;
+
+    this.contractType = contractType;
+  }
+  // Delete all events within the client
+  async clearState() {
+    this.createdContractEvents = [];
+  }
+
+  getAllCreatedContractEvents() {
```

Comment:
Do you think it might be nice to replace this (or have a separate method) that returns just a list of deployed addresses rather than the full event?

Reply:
Yes, but I don't really see a difference for now. If we find that we need this more (perhaps for the monitor) its easy enough to add the method. For now its just simpler to not do any filtering on the raw event data

Reply:
Perhaps we should rename this client to just a `FinancialContractFactoryClient` and we can use it to precompute some other contract data, such as `fundingRateIdentifier`, `collateralCurrency`, etc., most of which we do in the `funding-rate-proposer/proposer/src` code right now. WDYT?

Reply:
I see. So this is modeled after the other event clients. I could imagine it pulling down other contract metadata, but I could also imagine that being overkill for some/most of the other use cases.

The reason I think it could be nice to return a list of contract addresses is that that seems pretty universal. Put another way, if we wanted a liquidator bot to monitor _all_ emps, it would use this client for the same purpose: grabbing all addresses. Some other metadata may be useful in some cases, but certainly the fundingRateIdentifier wouldn't be needed by the liquidator bot, for example, and it might be a waste to query for that.

Still not strongly opinionated -- just a thought about generalizability.

Reply:
Nothing necessarily needs to be changed here. Just want to make sure you see this thread.

Reply:
> I see. So this is modeled after the other event clients. I could imagine it pulling down other contract metadata, but I could also imagine that being overkill for some/most of the other use cases.

Yeah problem would be if a user just wants contract addresses and no state, precomputing the state would slow the address query down.

I ultimately agree with you that this methid would be useful to return all addresses, so I'll rename this client to signify that its not just an event client, and add the method.

Reply:
Done.